File: Problems 0700-0799/735. Asteroid Collision/solution.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for asteroid in asteroids:
            if asteroid >= 0 or len(stack) == 0:
                stack.append(asteroid)
            else:
                prev = stack.pop()
                size = abs(asteroid)
                if prev < 0:
                    # we won't hit same direction
                    stack.append(prev)
                    stack.append(asteroid)
                elif prev < size:
                    # pop asteroids that explode because
                    # curr is bigger
                    while stack and stack[-1] >= 0 and stack[-1] < size:
                        stack.pop()
                    if stack and stack[-1] == size:
                        # both explode
                        stack.pop()
                    # only if asteroid beats every asteroid in stack
                    # does it survive or same direction asteroid
                    elif len(stack) == 0 or stack[-1] < 0:
                        stack.append(asteroid)

                elif prev > size:
                    # beats current asteroid
                    stack.append(prev)
                
        return stack

File: Problems 0700-0799/735. Asteroid Collision/solution.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> res;
        for (int asteroid : asteroids) {
            if (res.empty() || res[res.size()-1] < 0 == asteroid < 0 || res[res.size()-1] < 0) {
                res.push_back(asteroid);
            } else {
                // look for collisions (smaller)
                while (!res.empty() && res[res.size()-1] > 0 && res[res.size()-1] < abs(asteroid)) {
                    res.pop_back();
                }

                // same size collision (don't add asteroid because
                // it's exploded)
                if (!res.empty() && res[res.size()-1] > 0 && abs(res[res.size()-1]) == abs(asteroid)) {
                    res.pop_back();
                } else if (res.empty() || res[res.size()-1] < 0 == asteroid < 0) {
                    res.push_back(asteroid);
                }
            }
        }
        return res;
    }
};